server/tserver/src/test/java/org/apache/accumulo/tserver/tablet/RingBufferTest.java

```diff
@@ -0,0 +1,209 @@
+            startLock.wait(3000);
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+        long start = System.currentTimeMillis();
+        while (System.currentTimeMillis() - start < 2000) {
+          if (random.nextBoolean()) {
+            ring.add(UUID.randomUUID().toString());
+          } else {
+            ring.clear();
+          }
+        }
+      }
+    });
+    writerThread.setUncaughtExceptionHandler(handler);
+
+    // create a writer threads
+    Thread readerThread = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        synchronized (startLock) {
+          try {
+            ready.incrementAndGet();
+            startLock.wait(3000);
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+        long start = System.currentTimeMillis();
+        while (System.currentTimeMillis() - start < 2000) {
+          ring.toListNoFail();
+        }
+      }
+    });
+    readerThread.setUncaughtExceptionHandler(handler);
+
+    // start the threads
+    writerThread.start();
+    readerThread.start();
+
+    // wait until they are both waiting on the start lock
+    while (ready.get() < 2) {
+      Thread.sleep(100);
+    }
+
+    // unlock the threads
+    synchronized (startLock) {
+      ready.incrementAndGet();
+      startLock.notifyAll();
+    }
+
+    // wait for the threads to complete
+    while (writerThread.isAlive() || readerThread.isAlive()) {
+      Thread.sleep(100);
+    }
+
+    // ensure no exceptions were thrown
+    assertNull(handler.thrown, "Exception was thrown: " + handler.thrown);
+  }
+
+  @Test
+  public void maxAllowedSizeFails() {
+    try {
+      TabletTransactionLog.Ring<Integer> ring =
+          new TabletTransactionLog.Ring<>(TabletTransactionLog.Ring.MAX_ALLOWED_SIZE + 1);
+      fail("Expected failure creating ring larger than "
+          + TabletTransactionLog.Ring.MAX_ALLOWED_SIZE);
+    } catch (IllegalArgumentException e) {
+      // expected;
+    }
+  }
+
+  @Test
+  public void testOverrun() {
+    TabletTransactionLog.Ring<Integer> ring = new TabletTransactionLog.Ring<>(10);
+    for (int i = 0; i < TabletTransactionLog.Ring.OVERRUN_THRESHOLD; i++) {
+      ring.add(1);
+    }
+    assertEquals(10, ring.size());
+    assertEquals(TabletTransactionLog.Ring.OVERRUN_THRESHOLD - 1, ring.last());
+    assertEquals(TabletTransactionLog.Ring.OVERRUN_THRESHOLD - 10, ring.first());
+
+    ring.add(1);
+
+    assertEquals(10, ring.size());
+    assertEquals(TabletTransactionLog.Ring.OVERRUN_THRESHOLD, ring.last());
+    assertEquals(TabletTransactionLog.Ring.OVERRUN_THRESHOLD - 9, ring.first());
+
+    ring.add(1);
+
+    assertEquals(10, ring.size());
+    assertEquals(9, ring.last());
+    assertEquals(0, ring.first());
+
+    ring.add(1);
+
+    assertEquals(10, ring.size());
+    assertEquals(10, ring.last());
+    assertEquals(1, ring.first());
+  }
```

Comment:
I tried making this change locally and test did not pass.  It get past the first few calls of ring.toList and then eventually one fails after an addition.

```suggestion
  public void testOverrun() {
    TabletTransactionLog.Ring<Integer> ring = new TabletTransactionLog.Ring<>(10);
    int count = 0;
    for (int i = 0; i < TabletTransactionLog.Ring.OVERRUN_THRESHOLD; i++) {
      ring.add(count++);
    }

    assertEquals(IntStream.range(count-10, count).boxed().collect(Collectors.toList()), ring.toList());
    assertEquals(10, ring.size());
    assertEquals(TabletTransactionLog.Ring.OVERRUN_THRESHOLD - 1, ring.last());
    assertEquals(TabletTransactionLog.Ring.OVERRUN_THRESHOLD - 10, ring.first());

    ring.add(count++);

    assertEquals(IntStream.range(count-10, count).boxed().collect(Collectors.toList()), ring.toList());
    assertEquals(10, ring.size());
    assertEquals(TabletTransactionLog.Ring.OVERRUN_THRESHOLD, ring.last());
    assertEquals(TabletTransactionLog.Ring.OVERRUN_THRESHOLD - 9, ring.first());

    ring.add(count++);

    assertEquals(IntStream.range(count-10, count).boxed().collect(Collectors.toList()), ring.toList());
    assertEquals(10, ring.size());
    assertEquals(9, ring.last());
    assertEquals(0, ring.first());

    ring.add(count++);

    assertEquals(IntStream.range(count-10, count).boxed().collect(Collectors.toList()), ring.toList());
    assertEquals(10, ring.size());
    assertEquals(10, ring.last());
    assertEquals(1, ring.first());
  }
```

Reply:
oh of course, the overrun method is not keeping the correct position in the ring.  Good catch!!!!  Thank you!